ansible/roles/generate_blocks/tasks/main.yml

```diff
@@ -1,20 +1,71 @@
 ---
 
+# Init
+
 - name: Fail if vars are not set properly
   ansible.builtin.fail:
-    msg: 'num_blocks and balance_needed must always be set. set one of them to a non-zero value and the other one to zero'
-  when: num_blocks is not defined or balance_needed is not defined or (num_blocks != 0 and balance_needed != 0) or (num_blocks == 0 and balance_needed == 0)
+    msg: 'num_blocks, balance_needed or txid_list must always be set according to the value of generate'
+  when: >
+    (generate == 'blocks' and num_blocks is not defined) or
+    (generate == 'balance' and target_balance is not defined) or
+    (generate == 'confirmations' and (txid_list is not defined or tx_source is not defined))
 
 - name: Specify which networks permit generating blocks
   ansible.builtin.set_fact:
     generate_networks:
       - regtest
       - devnet
 
-- name: Generate blocks
-  ansible.builtin.include_tasks: generate_blocks.yml
+- name: Get start block count
+  ansible.builtin.command: dash-cli getblockcount
+  register: start_block_count
+  changed_when: false
+  failed_when: start_block_count.rc != 0
+
+# Start mining
+
+- name: Start generating blocks
+  ansible.builtin.include_role:
+    name: dashd_generate_miner
   when: dash_network in generate_networks
 
-- name: Wait for blocks
-  ansible.builtin.include_tasks: wait_for_blocks.yml
-  when: dash_network not in generate_networks
+# Option: blocks
+
+- name: Wait for {{ num_blocks ~ ' block(s)' }}
+  ansible.builtin.command: dash-cli getblockcount
+  when: generate == 'blocks'
+  register: getblockcount_result
+  until: getblockcount_result.stdout | int >= start_block_count.stdout | int + num_blocks | int
+  retries: 200
```

Comment:
Shouldn't this value depend on how many blocks / dash we generate?

Reply:
Good point, but I think we would need to know the current block subsidy in order to calculate this? 

I've never encountered an insufficient balance condition because we mine 4032 blocks with minimum difficulty on all networks anyway, so I don't think it really matters.

Reply:
Ok, so if you think it will never happen in any possible use cases you can keep like this I guess